Block: Code-title component.

StevenDufresne opened the pull request:
Related #162 

This PR moves and simplifies code related to the display line of code. 

| Type | |
| --- | --- |
| Method | <img width="666" alt="Screen Shot 2023-01-19 at 11 24 19 AM" src="https://user-images.githubusercontent.com/1657336/213340756-0c6a9b7b-61d3-4f19-865a-d2a908a7d4ed.png"> |
| Function | <img width="668" alt="Screen Shot 2023-01-19 at 11 23 45 AM" src="https://user-images.githubusercontent.com/1657336/213340765-6b0d0944-fe00-45ea-be1f-6839a1e5cef9.png"> |
| Hook | <img width="356" alt="Screen Shot 2023-01-19 at 11 23 21 AM" src="https://user-images.githubusercontent.com/1657336/213340772-984132ec-1513-499c-ad4f-9cd410105918.png"> |
| Class | <img width="446" alt="Screen Shot 2023-01-19 at 11 28 25 AM" src="https://user-images.githubusercontent.com/1657336/213341245-7315ca90-d9c7-450a-a704-80273d19da7a.png"> |


## Block Editor
I added the ability to control the container tag because this is also needed for rendering hooks.

<img width="280" alt="Screen Shot 2023-01-19 at 11 38 46 AM" src="https://user-images.githubusercontent.com/1657336/213342857-9e17e3d4-2911-40b4-9f11-e36535f9c51c.png">


## Testing
1. Load some single pages:
   - [wp_posts_list_table/]( http://localhost:8888/reference/classes/wp_posts_list_table/)
   - [wp_widget_rss/update/]( http://localhost:8888/reference/classes/wp_widget_rss/update/)
   - [add_menu_page/](http://localhost:8888/reference/functions/add_menu_page/)

StevenDufresne committed:
Move code to the code-title component.

StevenDufresne committed:
Allow the configuration of the tag.

StevenDufresne committed:
Escape the attribute.

adamwoodnz commented on source/wp-content/themes/wporg-developer-2023/src/code-title/style.scss:
```diff
@@ -1 +1,39 @@
+.wp-block-wporg-code-reference-title {
+	> * {
+		font-size: var(--wp--preset--font-size--extra-large);
+		line-height: 42px;
```
Best practice to use unitless, see https://developer.mozilla.org/en-US/docs/Web/CSS/line-height#values
```suggestion
		line-height: 1.73;
```

(value based on Figma)

adamwoodnz commented on source/wp-content/themes/wporg-developer-2023/src/code-title/style.scss:
```diff
@@ -1 +1,39 @@
+	> * {
+		font-size: var(--wp--preset--font-size--extra-large);
+		line-height: 42px;
+		color: #008a20;
```
I can see these colors are also used for code examples; is it worth adding presets for them, or sharing some other way so that we only have to change them in one place?

ryelle replied:
I agree these should be defined somewhere specific.
We can't add presets to `theme.json` without duplicating all the existing values, but we could sorta fake it in the CSS by creating some new custom properties. The only drawback is that they won't be available in the editor itself.

ryelle commented on source/wp-content/themes/wporg-developer-2023/src/code-title/style.scss:
```diff
@@ -1 +1,39 @@
+
+	.arg-type {
+		color: #d63638;
+		font-style: italic;
```
Looks like this is not italic in the design:

<img width="605" alt="Screenshot 2023-01-20 at 1 38 25 PM" src="https://user-images.githubusercontent.com/541093/213779693-a0f5bf11-650a-405a-8338-a71ba72834c1.png">

(though the italic style is fancy)

ryelle committed:
Add an `isLink` attribute and use it to render a link wrapper

ryelle committed:
Import cleanup

ryelle commented:
I added the `postId` context to the block.json so that the value would be passed in to the render function & we can use that -- it's just a little safer than relying on the global object for if/when we use this in a Query Loop or other nested context.

I also added an `isLink` property since otherwise the link was never being rendered. You can test it by manually setting a link on `<!-- wp:wporg/code-reference-title {"isLink":true} /-->`.

ryelle committed:
Add a ToggleControl for the link attribute

ryelle committed:
Remove the extra div wrapper

ryelle committed:
Add spacing support and add blockProps to the edit component

ryelle committed:
Clean up CSS

Use a unitless line-height, remove italics to match design, remove unnecessary comment.

ryelle committed:
Try local color palette for syntax colors

ryelle commented:
So I'm not sure the best approach in general, but it looks like the lack of `useBlockProps` in the edit function breaks the block in the editor… but including it _definitely_ applies the class names & styles twice.

<img width="715" alt="Screenshot 2023-01-20 at 4 35 21 PM" src="https://user-images.githubusercontent.com/541093/213809675-9d064d1f-1603-4366-8eb9-8852fd89f033.png">

After trying a janky workaround with checking `REST_REQUEST` in render, I looked at how the core blocks handle this and found that `ServerSideRender` has a prop for `skipBlockSupportAttributes` which solves the problem 😑 

So I've also enabled the dimension support for this block, and tidied up the output & CSS. By not nesting the element in a div wrapper, we can apply the font size directly to it, and when custom font sizes are used it just works.

Lastly, I tried a local palette for the syntax colors (see https://github.com/WordPress/wporg-developer/pull/171/commits/cf8bc8f70463c88b1b1417b6a43a7fa4cd763cac), which does generate these custom properties (on the `.wp-block-wporg-code-reference-title` class). In this way, they're not reusable across blocks, but it's a step in that direction. This `palette` value overrides the base theme one, so if this block had a color control you would only see these 5 colors available (no more blueberry, charcoal, etc). I don't really know if that's any better than defining them in CSS and not having them in the editor? So we can iterate there.

StevenDufresne commented:
Thanks for all the help!

StevenDufresne commented on source/wp-content/themes/wporg-developer-2023/src/code-title/block.php:
```diff
@@ -24,14 +24,33 @@ function init() {
+	}
+
+	$post_ID = $block->context['postId'];
+	$content = get_signature( $post_ID );
```
Any idea on how we will be able to render this dynamic block inside another dynamic block with the right context? In the hooks component, we'll need to do that. We'll need to loop through the hooks and construct a couple of these blocks.

I can't seem to find a way to retrieve a string of a _rendered_ dynamic block without using `do_blocks` which outputs the block to the document. 

My only other thought is to pass the `postId` through.

StevenDufresne replied:
Actually, disregard this, it appears like I had a typo that was causing the immediate output. 👍 `do_blocks` doesn't output itself :).
